In [13]:
using static Microsoft.DotNet.Interactive.Formatting.PocketViewTags;
using Microsoft.DotNet.Interactive.Formatting;
using Microsoft.DotNet.Interactive;
//using XPlot.Plotly.Interactive;
using System.Collections.Generic;
using System.IO;
using System;
using System.Text.RegularExpressions;

In [14]:
var fileExists = File.Exists("input.txt");
if (fileExists)
{
    display("This file exists.  Nothing has blown up.....yet.");
}

This file exists.  Nothing has blown up.....yet.

In [15]:
string[] lines = System.IO.File.ReadAllLines("input.txt");
lines.Length
//lines[0].Length


140

In [99]:
public class CharacterGrid
{
    public char[,] grid; // Internal 2D array for storing characters
    public int rows; // Number of rows in the grid
    public int columns; // Number of columns in the grid

    public CharacterGrid(int rows, int columns)
    {
        this.rows = rows;
        this.columns = columns;
        this.grid = new char[rows, columns];
    }

    // Get the character at a specific row and column
    public char GetCharacter(int row, int column)
    {
        if (!IsValidIndex(row, column))
        {
            return '.'; // return a useless '.' character if the index isn't valid
        }

        return grid[row, column];
    }

    // Set the character at a specific row and column
    public void SetCharacter(int row, int column, char character)
    {
        if (!IsValidIndex(row, column))
        {
            throw new IndexOutOfRangeException("Invalid row or column index");
        }

        grid[row, column] = character;
    }

    public void SetRow(string line, int rowNum)
    {
        int i = 0;
        foreach(char c in line)
        {
            SetCharacter(rowNum, i, c);
            i++;
        }
    }

    public List<PartNumber> GetUnverifiedParts(string line, int rowNum)
    {
        List<PartNumber> unverifiedParts = new List<PartNumber>();

        Regex regex = new Regex(@"(-?\d+)");

        foreach (Match match in regex.Matches(line))
        {
            int number = int.Parse(match.Groups[1].Value);
            int startPos = match.Index;
            int endPos = startPos + match.Length;
            PartNumber tempPart = new PartNumber(number, startPos, endPos, rowNum);
            unverifiedParts.Add(tempPart);
            //Console.WriteLine($"Line: {rowNum} Number: {number}, Start: {startPos}, End: {endPos}");
        }
        return unverifiedParts;
    }

    public List<PartNumber> GetVerifiedParts(List<PartNumber> unverifiedParts)
    {
        List<PartNumber> VerifiedParts = new List<PartNumber>();

        foreach (PartNumber pn in unverifiedParts)
        {
            bool isSpecial = false;
            foreach ((int row, int column) bc in pn.borderCharacters)
            {
                bool actuallyIsSpecial = IsSpecialCharacter(bc.row, bc.column);
                if (actuallyIsSpecial)
                {
                    isSpecial = true;
                }
            }
            if(isSpecial)
            {
                pn.isValid = true;
                VerifiedParts.Add(pn);
            }
        }
        return VerifiedParts;
    }

    public bool IsSpecialCharacter(int row, int column)
    {
        // looking for *, =, -, +, &, #, %, /, @, $
        char c = GetCharacter(row, column);
        bool special =  false;
        switch(c)
        {
            case '*':
            case '=':
            case '-':
            case '+':
            case '&':
            case '#':
            case '%':
            case '/':
            case '@':
            case '$':
                //Console.Out.WriteLine("Character " + c + " is special");
                special = true;
                break;
            default:
                special = false;
                break;
        }
        return special;

    }

    // Check if a row and column index are within the grid bounds
    public bool IsValidIndex(int row, int column)
    {
        return row >= 0 && row < rows && column >= 0 && column < columns;
    }

    public bool HasValidIndicator()
    {
        return false;
    }


    
}

public class PartNumber
{
    public int value;
    public int startColumn;
    public int endColumn;
    public int row;
    public bool isValid;
    public List<(int row, int column)> borderCharacters;

    public PartNumber(int value, int startColumn, int endColumn, int row)
    {
        this.value = value;
        this.startColumn = startColumn;
        this.endColumn = endColumn;
        this.row = row;
        borderCharacters = GetBorderCharacters();
    }

    public List<(int row, int column)> GetBorderCharacters()
    {
        List<(int row, int column)> borderChars = new List<(int row, int column)>();
        int leftColumn = this.startColumn - 1;
        int rightColumn = this.endColumn + 1;
        int topRow = this.row - 1;
        int bottomRow = this.row + 1;
        //Console.WriteLine($"TopRow: {topRow} BottomRow: {bottomRow}, LeftColumn: {leftColumn}, RightColumn: {rightColumn}");
        // Example - TopRow: 0 BottomRow: 2, LeftColumn: 2, RightColumn: 6
        // Add top and bottom rows.
        for (int i = leftColumn; i<=rightColumn; i++)
        {
            (int row, int column) temp = (topRow, i);
            (int row, int column) temp2 = (bottomRow, i);
            borderChars.Add(temp);
            borderChars.Add(temp2);
        }
        // Add middle row. Only add leftmost and rightmost columns.
        borderChars.Add((this.row, leftColumn));
        borderChars.Add((this.row, rightColumn));




        return borderChars;
    }

}



In [50]:
// Let's get a list of all characters, to specify the "special" ones.
List<char> allChars = lines.SelectMany(s => s.ToCharArray()).Distinct().ToList();

Console.WriteLine(string.Join(", ", allChars));

4, 8, ., 5, 0, 1, 3, 6, 2, 7, 9, *, =, -, +, &, #, %, /, @, $


In [101]:
CharacterGrid cg = new CharacterGrid(140, 140);
int i = 0;
List<PartNumber> parts = new List<PartNumber>();
foreach(string line in lines)
{
    cg.SetRow(line, i);
    List<PartNumber> tempParts = cg.GetUnverifiedParts(line, i);
    parts.AddRange(tempParts);
    i++;
}


//char x = cg.GetCharacter(137,136);

//bool isSpecial = cg.IsSpecialCharacter(137,136);

//PartNumber p = new PartNumber(48, 0, 1, 0);
//PartNumber p2 = new PartNumber(491,3, 5, 1);

//List<(int row, int column)> d = p2.GetBorderCharacters();

List<PartNumber> validParts = cg.GetVerifiedParts(parts);

int sum = 0;
foreach(PartNumber p in validParts)
{
    sum += p.value;
}
sum


491927

This is wrong. However, I'm not sure what will make it right. It says that my answer (491927) is too low. I assume that my search for special characters in the border characters, so I'll need to debug that.